kernel/arch/i386/start.S

```diff
@@ -23,7 +23,7 @@
                               MULTIBOOT_MEMORY_INFO |  \
                               MULTIBOOT_VIDEO_MODE)
 
-#define PAGE_DIR_PADDR ((offset page_size_buf) - KERNEL_BASE_VA)
+#define PAGE_DIR_PADDR ((offset page_size_buf) - KERNEL_VADDR + KERNEL_PADDR)
```

Comment:
No... you're using KERNEL_VADDR in too many place. ALWAYS anywhere KERNEL_VADDR should be replaced with `BASE_VA` simply because we don't care in all of these places the address of kernel's text! We care about the linear mapping of the physical memory in the virtual address space.